geometry3d/volLocalMax.cpp

```diff
@@ -0,0 +1,144 @@
+ * @author Bertrand Kerautret (\c bertrand.kerautret@loria.fr)
+ *
+ * @date 2015/12/22
+ *
+ *
+ * This file is part of the project DGtal-Tools-Contrib of the DGtal library.
+ */
+
+
+///////////////////////////////////////////////////////////////////////////////
+#include <iostream>
+#include "DGtal/base/Common.h"
+#include "DGtal/helpers/StdDefs.h"
+#include "DGtal/io/readers/GenericReader.h"
+#include "DGtal/io/boards/Board3D.h"
+#include "DGtal/io/DrawWithDisplay3DModifier.h"
+#include "DGtal/shapes/implicit/ImplicitBall.h"
+#include "DGTal/shapes/GaussDigitizer.h"
+
+
+#include <boost/program_options/options_description.hpp>
+#include <boost/program_options/parsers.hpp>
+#include <boost/program_options/variables_map.hpp>
+
+using namespace std;
+using namespace DGtal;
+using namespace Z3i;
+
+///////////////////////////////////////////////////////////////////////////////
+namespace po = boost::program_options;
+
+typedef ImageContainerBySTLVector<Z3i::Domain, unsigned char> Image3D;
+
+
+
+template<typename TImage>
+bool
+isLocalMax(const TImage &anImage, const Z3i::Point &aPoint, double aRadius){
+  typedef DGtal::ImplicitBall<DGtal::Z3i::Space> EuclideanBall;
+  typedef DGtal::GaussDigitizer<DGtal::Z3i::Space, EuclideanBall> DigitalShape;
+  EuclideanBall aBall (DGtal::Z3i::Point(aPoint[0], aPoint[1], aPoint[2]), aRadius);
+  unsigned int myVal = anImage(aPoint);
+  DigitalShape  gaussDig;
+  gaussDig.attach (aBall);
+  gaussDig.init(DGtal::Z3i::Point(aPoint[0]-(int)aRadius-1, aPoint[1]-(int)aRadius-1, aPoint[2]-(int)aRadius-1),
+                DGtal::Z3i::Point(aPoint[0]+(int)aRadius+1, aPoint[1]+(int)aRadius+1, aPoint[2]+(int)aRadius+1), 1);
+  DGtal::Z3i::Domain dom = gaussDig.getDomain();
+  for( DGtal::Z3i::Domain::ConstIterator it = dom.begin(); it!=dom.end(); it++){
+    if( *it != aPoint){
+      if(gaussDig(*it) && anImage.domain().isInside(*it) ){
+        if (myVal <= anImage(*it) ) {
+          return false;
+        }
+      }
+    }
+  }
+  return  true;
+}
+
+
+
+
+
+int main( int argc, char** argv )
+{
+  // parse command line ----------------------------------------------
+  po::options_description general_opt("Allowed options are: ");
+  general_opt.add_options()
+  ("help,h", "display this message")
+  ("input,i", po::value<std::string>(), "vol file (.vol) , pgm3d (.p3d or .pgm3d, pgm (with 3 dims)) file." )
+  ("output,o", po::value<std::string>(), "Output SDP filename" )
+  ("ballSize,b",  po::value<double>()->default_value(3.0), "set the ball size." );
+  
+  
+  bool parseOK=true;
+  po::variables_map vm;
+  try
+  {
+    po::store(po::parse_command_line(argc, argv, general_opt), vm);
+  } catch(const std::exception& ex)
+  {
+    parseOK=false;
+    trace.info()<< "Error checking program options: "<< ex.what()<< endl;
+  }
+  po::notify(vm);
+  if( !parseOK || vm.count("help")||argc<=1)
+  {
+    std::cout << "Usage: " << argv[0] << " [input-file]\n" << "Extraction the local maximal of a vol images. \n"
```

Comment:
extraction of local maxima of a vol image within a spherical kernel of radius '--ballSize'   ?